onnxruntime/core/providers/cuda/cuda_execution_provider.cc

```diff
@@ -407,7 +407,14 @@ Status CUDAExecutionProvider::EnqueueDeferredRelease() {
     cpu_buffers_info->buffers = buffers;
     // Release the ownership of cpu_buffers_info so that the underlying
     // object will keep alive until the end of ReleaseCpuBufferCallback.
-    CUDA_RETURN_IF_ERROR(cudaLaunchHostFunc(stream, ReleaseCpuBufferCallback, cpu_buffers_info.release()));
+    if (cpu_buffers_info->allocator->Info().alloc_type == OrtArenaAllocator) {
+      // Release memory asynchronously to avoid blocking the compute stream.
+      CUDA_RETURN_IF_ERROR(cudaLaunchHostFunc(stream, ReleaseCpuBufferCallback, cpu_buffers_info.release()));
+    } else {
+      // cudaFreeHost and all other CUDA APIs cannot be called by cudaLaunchHostFunc per spec.
+      // So we just do synchrous release.
+      ReleaseCpuBufferCallback(cpu_buffers_info.release());
```

Comment:
should we perform a stream sync before release it?

Reply:
Yeah, you're right. I check [here ](https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#implicit-synchronization)and `cudaFreeHost` doesn't block the stream.